_posts/2019-06-05-uncertainty-estimates-tfprobability/uncertainty_tfprobability.Rmd

```diff
@@ -0,0 +1,782 @@
+knitr::include_graphics("images/uci_aleatoric.png")
+```
+
+This looks pretty good we'd say! How about epistemic uncertainty?
+
+### Single predictor: Epistemic uncertainty
+
+Here's the code:
+
+```{r}
+posterior_mean_field <-
+  function(kernel_size,
+           bias_size = 0,
+           dtype = NULL) {
+    n <- kernel_size + bias_size
+    c <- log(expm1(1))
+    keras_model_sequential(list(
+      layer_variable(shape = 2 * n, dtype = dtype),
+      layer_distribution_lambda(
+        make_distribution_fn = function(t) {
+          tfd_independent(tfd_normal(
+            loc = t[1:n],
+            scale = 1e-5 + tf$nn$softplus(c + t[(n + 1):(2 * n)])
+          ), reinterpreted_batch_ndims = 1)
+        }
+      )
+    ))
+  }
+
+prior_trainable <-
+  function(kernel_size,
+           bias_size = 0,
+           dtype = NULL) {
+    n <- kernel_size + bias_size
+    keras_model_sequential() %>%
+      layer_variable(n, dtype = dtype, trainable = TRUE) %>%
+      layer_distribution_lambda(function(t) {
+        tfd_independent(tfd_normal(loc = t, scale = 1),
+                        reinterpreted_batch_ndims = 1)
+      })
+  }
+
+model <- keras_model_sequential() %>%
+  layer_dense_variational(
+    units = 1,
+    make_posterior_fn = posterior_mean_field,
+    make_prior_fn = prior_trainable,
+    kl_weight = 1 / n,
+    activation = "linear",
+  ) %>%
+  layer_distribution_lambda(function(x)
+    tfd_normal(loc = x, scale = 1))
+
+negloglik <- function(y, model) - (model %>% tfd_log_prob(y))
+model %>% compile(optimizer = optimizer_adam(lr = learning_rate), loss = negloglik)
+hist <-
+  model %>% fit(
+    X_train[, i, drop = FALSE],
+    y_train,
+    validation_data = list(X_val[, i, drop = FALSE], y_val),
+    epochs = n_epochs,
+    batch_size = batch_size
+  )
+
+yhats <- purrr::map(1:100, function(x)
+  yhat <- model(tf$constant(X_val[, i, drop = FALSE])))
+  
+means <-
+  purrr::map(yhats, purrr::compose(as.matrix, tfd_mean)) %>% abind::abind()
+
+lines <- data.frame(cbind(X_val[, i], means)) %>%
+  gather(key = run, value = value,-X1)
+
+mean <- apply(means, 1, mean)
+ggplot(data.frame(x = X_val[, i], y = y_val, mean = as.numeric(mean)), aes(x, y)) +
+  geom_point() +
+  geom_line(aes(x = X_val[, i], y = mean), color = "violet", size = 1.5) +
+  geom_line(
+    data = lines,
+    aes(x = X1, y = value, color = run),
+    alpha = 0.3,
+    size = 0.5
+  ) +
+  theme(legend.position = "none")
+```
+
+And this is the result. 
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset", fig.cap = "Epistemic uncertainty on the Combined Cycle Power Plant Data Set; single predictors."}
+knitr::include_graphics("images/uci_epistemic.png")
+```
+
+As with the simulated data, the linear models seems to "do the right thing". And here too, we think we will want to augment this with the spread in the data: Thus, on to way 3.
```

Comment:
> Thus, on to way 3.

I would write out the word three, here.

Reply:
oh, right, trying to do that consistently otherwise...